### ISSUE TITLE
Config to automatically increment to next major, minor or patch version

### ISSUE THREAD

**Issue by mkurz** [sbt/sbt-dynver#227]
I understand how sbt-dynver works and how it comes up with the version string.

However IMHO it would be usefull if there would be a config which increments either the major, minor or patch version, plus ignore the distance.
For example, right now in Play's master branch we are working towards 2.9.0, also we are in the process of moving to GitHub actions and sbt-ci-release and want to publish sonatype snapshots on each push to the master branch.
Right now the version in Play's master branch looks like this:
* `2.8.1+1506-a4b4198b-SNAPSHOT`
* `2.8.1+1506-a4b4198b+20220214-1249-SNAPSHOT` when dirty

(2.8.1 because the 2.8.1 was tagged before we branched off a 2.8.x branch).
So people asked us already why this is 2.8.1... and not 2.9.0-SNAPSHOT.
So actually what we want to publish to the sonatype repo is:
* `2.9.0-a4b4198b-SNAPSHOT`
* `2.9.0-a4b4198b+20220214-1249-SNAPSHOT` when dirty

which makes it clearer these are snapshots of the upcoming 2.9.0 release. I think this is what people are used to.
So actually we would have to remove the distance and increment one of the major/minor/patch parts.
The same for the 2.8.x branch right now, it would be nice to publish `2.8.14-...-SNAPSHOT` versions instead of `2.8.13+42-....-SNAPSHOT`.

Actually we have a pull request already which achieves that, however I think it would be nice to have this build into sbt-dynver: https://github.com/playframework/playframework/pull/11168

**Comment by dwijnand** [comment]
One of the goals of sbt-dynver is not require build changes, just rely on the git metadata.  What you're asking requires making a change to the build from using `increaseMinorVersion` to `increasePatchVersion` (or viceversa).

What I suggest instead of tagging the start of the new version series, for instance 2.9.0-M0.

**Comment by ihostage** [comment]
> What I suggest instead of tagging the start of the new version series, for instance 2.9.0-M0.

@dwijnand Sorry, Dale, but it's a not good suggestion 😞 Because it's not only about the new version series, it's about **all** versions. Having count of tags 2 times more than now is too much, IMHO.

> One of the goals of sbt-dynver is not require build changes, just rely on the git metadata. What you're asking requires making a change to the build from using `increaseMinorVersion` to `increasePatchVersion` (or viceversa).

Yep, but It's a very small **one-time** change for the versions series branch. I think that @mkurz proposal makes sense. For `sbt-dynver` it will be just a few helper functions that don't break current behavior and versions semantic but can be useful for someone who wants to use different semantics like in Play 😉 Let's be honest it's default version semantics for Maven repositories 🤷‍♂️ And how Matthias wrote, Play already has the draft of implementation of this semantics and she can stay in Play, but it looks like she can be implemented in `sbt-dynver` and useful not only for Play.

**Comment by dwijnand** [comment]
You're right, it wouldn't be just for new version series, and I can see how once per version series is a lot less work than once per release.

I sympathise with the desire, but I want to be careful adding complexity here.  So I may be convinced that this can be added to sbt-dynver is someone wants to send a PR with the code and README changes for this.

**Comment by mkurz** [comment]
@eed3si9n I think this issue should be re-opened, what do you think (also see last comment where Dale agrees this would make sense if implemented carefully)

**Comment by raquo** [comment]
Hi, I've been copy-pasting [this PlayFramework VersionHelper file](https://github.com/playframework/playframework/blob/main/project/VersionHelper.scala) between my projects – thanks for that, that helper works exactly how I need it.

Ideally it should be published somewhere to reduce the amount of boilerplate that library authors have to deal with to get the desired behaviour. If adding this version helper to sbt-dynver is controversial, can this perhaps be published as a separate plugin instead? It would still be better than the status quo.

**Comment by mkurz** [comment]
_Somehow_ related, but not really, still linking them here, maybe there is something useful there:
- https://github.com/sbt/sbt-release/issues/398
- https://github.com/sbt/sbt-release/pull/397
- https://github.com/sbt/sbt-release/pull/393

**Comment by SethTisue** [comment]
related: https://github.com/scalacenter/sbt-version-policy — it doesn't increment the version number for you, but it forces you to keep your intentions and your actions w/r/t versioning aligned

**Comment by SethTisue** [comment]
#76 is a duplicate or near-duplicate

**Comment by eed3si9n** [comment]
@SethTisue How so? It seems #76 is mostly about making sure that the version state is consistent with the underlying git state by running `versionCheck` etc, whereas this issue is more about creating a setting to influence a future-oriented SNAPSHOT version, i.e. `2.9.0-a4b4198b-SNAPSHOT` when the current tag is `2.8.1` etc?

**Comment by SethTisue** [comment]
I'd accept "related".
